Implement graph tool

nico-abram opened the pull request:
This pull request should implement the graph tool

michaelforney commented on tool.c:
```diff
@@ -171,6 +171,72 @@ commands(int argc, char *argv[])
+graphviznode(struct node *n)
+{
+	size_t i;
+	char *order_only;
```
I'd make this `const char *`, and maybe call it `style`.

michaelforney commented on tool.c:
```diff
@@ -171,6 +171,72 @@ commands(int argc, char *argv[])
+		for (i = 0; i < e->nin; i++) {
+			order_only = "";
+			if (i >= e->inorderidx)
+				order_only = " style=dotted";
```
This seems like a good place to use the conditional operator:
```
order_only = i >= e->inorderidx ? " style=dotted" : "";
```

michaelforney commented on tool.c:
```diff
@@ -171,6 +171,72 @@ commands(int argc, char *argv[])
+	puts("}");
+
+	fflush(stdout);
+	if (ferror(stdout))
```
You could combine these lines to match the rest of the tools:
```
if (fflush(stdout) || ferror(stdout))
```

michaelforney commented on tool.c:
```diff
@@ -171,6 +171,72 @@ commands(int argc, char *argv[])
+		return;
+	e->flags |= FLAG_WORK;
+
+	for (i = 0; i < e->nin; i++)
```
Minor nitpick: use `++i` for consistency with the rest of the codebase.

michaelforney commented on tool.c:
```diff
@@ -171,6 +171,72 @@ commands(int argc, char *argv[])
+	puts("node [fontsize=10, shape=box, height=0.25]");
+	puts("edge [fontsize=10]");
+
+	if (argc) {
```
This needs to be `argc > 1`, since `argv[0]` is always the tool name.

michaelforney commented on tool.c:
```diff
@@ -171,6 +171,72 @@ commands(int argc, char *argv[])
+		graphviznode(e->in[i]);
+
+	if (e->nin == 1 && e->nout == 1) {
+		printf("\"%p\" -> \"%p\" [label=\" %s\"]\n", (void *)e->in[0], (void *)e->out[0], e->rule->name);
```
Do you know the reason for the leading space? Ninja says it's for aesthetics, but it seems to look fine to me without it.

nico-abram replied:
I don't, I was following ninja, looks fine without to me too. Removed it for now.

michaelforney commented on tool.c:
```diff
@@ -171,6 +171,72 @@ commands(int argc, char *argv[])
+}
+
+static int
+graphviz(int argc, char *argv[])
```
I'd call this `graph` (even though it is a bit generic), so that the function name always matches the tool name.

nico-abram replied:
Also renamed the recursive node function

michaelforney commented on tool.c:
```diff
@@ -387,6 +453,7 @@ static const struct tool tools[] = {
 	{"compdb", "dump compilation database", compdb},
 	{"query", "show incoming/outgoing edges for a path", query},
 	{"targets", "list targets", targets},
+	{"graph", "output graphviz dot file for targets", graphviz},
```
Could you sort this by tool name? Also, try to order the functions in the same order they appear in this list. I know `query` and `targets` are currently out of order, but I'll fix this after merging this PR.

michaelforney reviewed (commented):
This looks pretty good, thanks for working on it!

I have a few minor stylistic comments, but apart from those, we just need to document this tool in the manual and this should be good to go.

nico-abram commented:
Thanks for the review. Pushed a separate commit with the fixes to ease review (Although since I moved the functions I don't think it helps). I can squash if needed.

Also added the manual entry, apologies if I did it wrong, I've never touched man pages before.

michaelforney commented:
Looks good, merged. Thanks again!
